### ISSUE TITLE
install-app-deps: Configuring yargs through package.json is deprecated

### ISSUE THREAD

**Issue by ivancuric** [electron-userland/electron-builder#3751]
* **Version**: 20.39.0

* **Target**: 4.0.8 on platform=win32 arch=x64

I have a few local native modules linked via `file:` in `package.json`.

```
> electron-builder install-app-deps

Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
  • electron-builder version=20.39.0
  • loaded configuration file=package.json ("build" field)
  • rebuilding native production dependencies platform=win32 arch=x64
```

What are the yargs errors about?

**Comment by nathanlesage** [comment]
I just discovered them myself. It appears that this pertains to the command line arguments. For instance, what I've been doing in my `package.json` for the last year was put these commands in the `scripts` section:

```json
...
"scripts": {
        "postinstall": "electron-builder install-app-deps",
        "start": "electron .",
        "build:quick": "electron-builder --dir",
        "release:this": "electron-builder",
        "release:mac": "electron-builder --mac",
        "release:win": "electron-builder --win",
        "release:linux": "electron-builder --linux"
    },
```

And `yargs` is basically a JavaScript package that parses command line arguments.

So I'm pretty sure these warnings mean: "Please don't use `--dir`, `--mac` etc. anymore, but use the Javascript API directly!"

I.e.: Run electron-builder from within a node.js-script.

**Comment by ivancuric** [comment]
https://www.electron.build/api/electron-builder
I don't see any way to run it using the provided API.

**Comment by nathanlesage** [comment]
@ivancuric It's just weirdly documented.

As it says in the API docs:

![image](https://user-images.githubusercontent.com/17251683/54532521-37ad6080-4988-11e9-9da1-476666da04bf.png)

Raw Options refer to the CLI options and there you'll find this:

![image](https://user-images.githubusercontent.com/17251683/54532571-514ea800-4988-11e9-9617-bde6b4024683.png)

So CLI equals API arguments, i.e. you'd have to pass the command line arguments into the `package.json`s `build`-property and hope you place them at the right position in the tree.

-----

Downside of this is obviously that you'd need several different scripts to pass data to the builder so that it does not always trigger a full chain-rebuild if you just want to drop a testing app into the release directory, so you'd have to do something like this to avoid this:

```javascript
// Some script file, i.e. build.js
const builder = require('electron-builder')

// ... Some logic that builds up the build field, e.g.:
let options = {
  "win": {
    "target": [
      {
        "target": "nsis",
        "arch": [
          "x64",
          "ia32"
        ]
      }
    ]
  }

builder.build(options).then((sth) => {
  // I have literally no idea what would be passed
  // during a successful call, maybe just dump it
  // to the console
  console.log(sth)
}).catch((e) => {
  // Some error handling
  console.error(e)
})
```

**Comment by ivancuric** [comment]
Ah, thanks!

**Comment by mubaidr** [comment]
Thanks @nathanlesage 

By the way on successful build it returns paths to all the packages built.

**Comment by nathanlesage** [comment]
You're welcome! 🎉

And @mubaidr: classic.

**Comment by mubaidr** [comment]
Sample build script:

```js
const builder = require('electron-builder')
const Platform = builder.Platform

const config = {
  "win": {
    "target": [
      {
        target: ['nsis', 'zip', 'portable'],
        "arch": [
          "x64",
          "ia32"
        ]
      }
    ]
  }

builder
  .build({
    targets: Platform.WINDOWS.createTarget(),
    config,
  })
  .then(m => {
    console.log(m)
  })
  .catch(e => {
    console.error(e)
  })

```

**Comment by jgierer12** [comment]
This occurs for me even if I'm running `electron-builder` from the terminal, with and without options and even with things like `--help` and `--version` which are supposed to be run from the CLI.

**Comment by stvenyin** [comment]
yes,i'am
![image](https://user-images.githubusercontent.com/32395964/54731831-70894780-4bcb-11e9-832c-09557ec6de91.png)

**Comment by stvenyin** [comment]
configure "scripts": {
        "postinstall": "electron-builder install-app-deps",
        "start": "electron .",
        "build:quick": "electron-builder --dir",
        "release:this": "electron-builder",
        "release:mac": "electron-builder --mac",
        "release:win": "electron-builder --win",
        "release:linux": "electron-builder --linux"
    },
https://www.electron.build/api/electron-builder
Thanks!

**Comment by stvenyin** [comment]
![image](https://user-images.githubusercontent.com/32395964/54731960-26549600-4bcc-11e9-9d49-5c2f94854736.png)

**Comment by Faksprod** [comment]
Same problem here when I tried to run code containing in my package.json file from the CLI (Terminal) on MacOS Mojave (was working well 3 months ago before my yesterday Electron update).

- Electron: 4.1.3
- Electron-builder: 20.39.0

**package.json**

```
{
    "scripts": {
        "start": "electron .",
        "build": "electron-builder",
        "build-mac": "electron-builder --mac"
    },
    "build": {
        "appId": "com.myCompany.myApp",
        "productName": "myAppName",
        "copyright": "Copyright © 2019 myCompany",
        "mac":{
            "target":"mas",
            "type":"distribution",
            "provisioningProfile":"myApp.provisionprofile",
            "identity": "MyCompany (idNumber)",
        },
        "directories": {
            "output": "build"
        }
    }
}
```


When I try to run `npm run build-mac` it gives me this error message `Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.`

Thanks to @nathanlesage I finally found the solution. It wasn't really clear for me about where I should put the **build.js** file et how to execute it (I'm not a Node JS pro). If someone is in my case, here are more detailed steps:

- Create a new JS file in the folder containing the **package.json** file.
- You can give any name to this new JS file (eg: **myCustomScript.js**).
- In this **myCustomScript.js** file you will use the **electron-builder JS API** to run your build instead of using the CLI command `electron-builder --mac.`
- Put [this script (from the electron-builder doc)](https://www.electron.build/#programmatic-usage).
- Change the **targets** property if needed and custom the **config** property.
- In the **config** property you just have to copy paste the object contained in the "build" tree of your **package.json**. Then you can remove the "build" part from your **package.json** tree cause you don't need here anymore.
- Finally, to call and execute your **myCustomScript.js** you can add a line in the **scripts** property of your **package.json** like this:
```
"scripts": {
    "myCustomScript": "node myCustomScript.js"
}
```
- Then you can call it from the Terminal like this `npm run myCustomScript`

So now, my file looks like below and it works like a charm!

**package.json (new)**
```
{
    "scripts": {
        "start": "electron .",
        "build": "electron-builder",
        "myCustomScript": "node myCustomScript.js"
    }
}
```

**myCustomScript.js**
```
"use strict";

const builder = require("electron-builder");
const Platform = builder.Platform;

builder.build({
    targets: Platform.MAC.createTarget(),
    config: {
        
        "directories": {
            "output": "build"
        },
        "appId": "com.myCompany.myApp",
        "productName": "myAppName",
        "copyright": "Copyright © 2019 myCompany",

        "mac":{
            "target":"mas",
            "type":"distribution",
            "provisioningProfile":"myApp.provisionprofile",
            "identity": "MyCompany (idNumber)"
        }
    }
})
.then(() => {
    // handle result
    console.log('Build OK!');
})
.catch((error) => {
    // handle error
    console.log(error);
})
```

**Comment by Ancinek** [comment]
How about the `--publish` option? How should I apply this?

**Comment by firede** [comment]
> Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.

This warning is from `yargs`, not `electron-builder`, so it’s great to keep using the CLI.
I created a PR to fix this internal error: #3852
